Add batch payments

Aditya-A-G opened the pull request:
![Screenshot from 2023-06-15 17-19-05](https://github.com/optriment/web3-employment/assets/65503300/393965a5-bd1e-495b-b6bc-32028c367128)

gruz0 commented on next.config.js:
```diff
@@ -31,6 +31,7 @@ const nextConfig = {
     tokenAddress: process.env.TOKEN_ADDRESS,
     tokenDecimals: process.env.TOKEN_DECIMALS,
     tokenSymbol: process.env.TOKEN_SYMBOL,
+    batchContractAddress: process.env.BATCH_CONTRACT_ADDRESS,
```
Please add this variable to `.env.example`

Aditya-A-G commented on src/utils/batchTransfer/useBatchTransfer.ts:
```diff
@@ -0,0 +1,112 @@
+  const [isLoading, setIsLoading] = useState<boolean>(false)
+  const [error, setError] = useState<string>('')
+
+  useEffect(() => {
```
The useEffect block in useBatchTransfer is used to handle errors from the useBatchPayment hook. It listens for changes in the contractError state returned by useBatchPayment and updates the local error state in useBatchTransfer accordingly. This allows us to handle any errors that occur during the retrieval of batch payment contract data.

gruz0 commented on src/screens/groups/batch-payment-screen.tsx:
```diff
@@ -0,0 +1,251 @@
+      setIsLoading(true)
+
+      try {
+        const group = await api.getGroupById(groupId)
```
We should not list recipients without wallets.

There are two possible ways how to optimize such code:

1. Implement new function on API side to return only recipients with non-empty wallet
2. Filterize them on frontend side and keep `getGroupById` as it is right now

WDYT?

gruz0 replied:
For me 2 option is more easy to implement, but option 2 provides us a way to test such an API endpoint.

gruz0 replied:
It's not necessary right now, maybe we can move this discussion and create a new issue.

Aditya-A-G replied:
option 1 looks good to me. and yes, it's good to create a new issue for this

gruz0 replied:
Let's create an issue 👍

Aditya-A-G commented:
Changes Made:
- Implemented feature to disallow selecting people with zero amount in the last column.
- Updated the label on the "Send Payments" button to display the total amount before clicking, providing the exact price to the customer.
- Added "USDT" to the last column heading, indicating the currency.
- Moved the "comment" value below the recipient name.


![Screenshot from 2023-06-16 18-41-46](https://github.com/optriment/web3-employment/assets/65503300/4d8abc08-001e-4f7d-83f4-e55bb35acf01)

gruz0 commented on src/screens/groups/components/transaction-dialog/batch-transaction-dialog.tsx:
```diff
@@ -0,0 +1,114 @@
+      open={open}
+      size="tiny"
+    >
+      <Modal.Header>Send money to {' ' + groupName}</Modal.Header>
```
We don't need `{' ' + groupName}` here. Just use `{groupName}`

Aditya-A-G replied:
Hmm, i think prettier extension added it, i will remove it.

gruz0 commented on src/screens/groups/batch-payment-screen.tsx:
```diff
@@ -0,0 +1,275 @@
+
+    allIds?.forEach((id) => {
+      const amount = selectedRecipients[id]?.amount
+      if (amount !== 0) {
```
I think here is a bug, I need to test it carefully

Aditya-A-G replied:
yeahh sure! why i have added that condition is - let's say there are x recipients and one of them has amount 0, so now if user click's on select all button that recipient who's having amount 0 will not get selected but his amount value was getting resetting to default. so now if user again click's on select all button then as that recipient has his amount value more than 0, he will also get selected, i thought this is a wrong behaviour that amount value should not get automatically reset from 0 to xx without user doing it. so now to fix that bug that condition helps, if user's amount value is 0 that will not get reset to default value.

gruz0 commented on src/screens/groups/batch-payment-screen.tsx:
```diff
@@ -0,0 +1,275 @@
+
+  useEffect(() => {
+    if (data && data.recipients) {
+      const initialSelectedRecipients: SelectedRecipients = {}
```
Please add new line here

gruz0 commented on src/screens/groups/batch-payment-screen.tsx:
```diff
@@ -0,0 +1,275 @@
+          amount: recipient.salary ? recipient.salary : 0,
+          address: recipient.wallet_address ? recipient.wallet_address : '',
+        }
+      })
```
Please add new line here

gruz0 commented on src/screens/groups/batch-payment-screen.tsx:
```diff
@@ -0,0 +1,275 @@
+      (total, recipient) => {
+        if (recipient.selected) {
+          return total + recipient.amount
+        }
```
Please add new line here

gruz0 commented on src/screens/groups/components/recipients-list/batch-recipients-list.tsx:
```diff
@@ -0,0 +1,89 @@
+}: Props) => {
+  const handleRowSelect = (id: string) => {
+    const selectedAmount = selectedRecipients[id]?.amount
+    if (selectedAmount === 0) return // Disallow selection if the amount is zero
```
We don't need that comment, it's useless

gruz0 commented on src/screens/groups/components/recipients-list/batch-recipients-list.tsx:
```diff
@@ -0,0 +1,89 @@
+      </Table.Header>
+      <Table.Body>
+        {recipients.map((recipient) => (
+          <Table.Row warning={!!recipient.archived_at} key={recipient.id}>
```
Can you please extract this `<Table.Row>` into single file?

Aditya-A-G replied:
sure, i will

gruz0 commented on src/screens/groups/components/recipients-list/index.ts:
```diff
@@ -1 +1,2 @@
 export { default as RecipientsList } from './recipients-list'
+export { default as BatchRecipientsList } from './batch-recipients-list'
```
Frankly speaking, this component should be in separate folder, because there is no common behaviour between two different lists. And `batch-transaction-dialog` also.

Aditya-A-G replied:
shall i create two directories ? 1] batch-recipients-list 2] batch-transaction-dialog and then keeping their respective file in them, WDYT

gruz0 commented on src/screens/groups/components/batch-recipients-list/batch-recipient-info.tsx:
```diff
@@ -0,0 +1,56 @@
+  handleRowSelect,
+  handleAmountChange,
+}: Props) => (
+  <>
```
We can remove this fragment, because `<Table.Row>` wraps all the following text.

gruz0 commented on src/screens/groups/components/batch-recipients-list/batch-recipient-info.tsx:
```diff
@@ -0,0 +1,56 @@
+        <Input
+          type="number"
+          value={
+            selectedRecipients[recipient.id]?.amount !== undefined
```
These 3 lines of code look crazy...

First of all, we don't need to pass `selectedRecipients` through props to this component, because the only responsibility for such a component – is to render a specific row without any knowledge of other records.

You can do all required logic in parent component and pass only simplified object, like `selectedRecipient` (and keep `recipient` (dto) as it is.

gruz0 commented on src/screens/groups/components/batch-recipients-list/batch-recipients-list.tsx:
```diff
@@ -0,0 +1,66 @@
+    id: string,
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    const amount = event.target.value ? parseFloat(event.target.value) : null
```
I think there is another possible approach how to simplify this line:

```
const amount = parseFloat(event.target.value)
onAmountChange(id, finalAmount)
```

We don't need to create `amount` variable with `null` value and in the next line check this variable for null-ish value. So, I think we can simplify code with an example from above.

Aditya-A-G replied:
we need to set the value to 0 if `parseFloat(event.target.value)` is NaN
`const amount = parseFloat(event.target.value) || 0`
`onAmountChange(id, amount)`

gruz0 commented on src/screens/groups/components/batch-recipients-list/batch-recipients-list.tsx:
```diff
@@ -0,0 +1,66 @@
+}: Props) => {
+  const handleRowSelect = (id: string) => {
+    const selectedAmount = selectedRecipients[id]?.amount
+    if (selectedAmount === 0) return
```
We don't need this extra variable just to check condition. You can simplify:

```
if (parseInt(selectedRecipients[id]?.amount) <= 0) return

onRowSelect(id)
```

Did you check your code with negative numbers? Is that work?

Aditya-A-G replied:
yes it works, user can't enter negative values. i have set the minimum value to 0 
![Screenshot from 2023-06-20 13-04-53](https://github.com/optriment/web3-employment/assets/65503300/78a28cf2-51b3-461a-8540-e00b0821a016)

Aditya-A-G replied:
upon rechecking still i found a way how user can enter negative value so i have used Math.abs in handleAmountChange so the amount can never be negative 
`const amount = Math.abs(parseFloat(event.target.value)) || 0`
`onAmountChange(id, amount)`
